src/librustc_mir/interpret/intrinsics.rs

```diff
@@ -157,6 +157,23 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M>
                 let dest = self.force_allocation(dest)?;
                 self.copy_op(args[0], dest.into())?;
             }
+            "write_bytes" => {
+                let ptr = self.read_value(args[0])?;
+                let mplace = self.ref_to_mplace(ptr)?;
+                let val = self.read_scalar(args[1])?.to_u8()?;
+                let count = self.read_scalar(args[2])?.to_usize(&self.memory)?;
+                if mplace.layout.is_zst() || count == 0 {
+                    return Ok(true);
+                }
+                if let Some(byte_count) = mplace.layout.size.checked_mul(count, &self.memory) {
+                    self.memory.write_repeat(mplace.ptr, val, byte_count)?;
+                } else {
+                    return err!(Intrinsic(
+                        format!("Overflowing computing `count * size_of::<T>()` ({} * {}) in {}",
+                                count, mplace.layout.size.bytes(), intrinsic_name),
+                    ));
+                }
+            }
```

Comment:
Why is this code so different from https://github.com/solson/miri/blob/master/src/intrinsic.rs#L420 ?

Also note that the ZST optimization here is wrong. It is UB to call `write_bytes` on an unaligned pointer even for a ZST. Your code fails to check for that.